### ISSUE TITLE
Highlight Character, String or Regex

### ISSUE THREAD

**Issue by windymilla** [DistributedProofreaders/guiguts-py#41]
In GG1 Search menu

**Comment by tangledhelix** [comment]
I'm looking into this.

**Comment by tangledhelix** [comment]
Key bind note on this: in GG1 the keybind is defined as `Ctrl+$::altkeyname+h`, which on Mac comes out to Ctrl+Command+h; an unusual combination on Macs; investigate if another combo is open or would make more sense (for this and the other platforms). I assume this is Ctrl-Alt-h on Windows (would have to verify).

The problem noted in another ticket about the Opt key not seeming to work is an issue to look into further, it might help here. Although another note is that Cmd-Opt-h is consumed by the global "Hide Others" command, taken by the windowserver and not available to remap.

**Comment by tangledhelix** [comment]
That said, Ctrl-Cmd-h might work on Mac and be available, so it may be simplest to actually leave as-is. It just doesn't follow our GG2 convention.

**Comment by windymilla** [comment]
Happy if you decide not to allocate a key binding at this stage.

**Comment by tangledhelix** [comment]
During implementation I discovered the highlighting tools in GG1 don't behave as I believed they did.

## What I believed

"Highlight alignment column" is a persistent mode, turned on (or off) until user changes its state.

"Highlight surrounding quotes/brackets" is the same, on or off until the user changes the state.

The "manual" highlighter tools, when run, highlight the relevant things, but are mutually exclusive. So for example, "Highlight single quotes in selection" would do so, but if you then "Highlight double quotes in selection," first the single-quote highlight is removed before the new highlight is applied. Similarly, "Highlight string or regex" would remove those other highlights, then apply its highlight.

But while the "manual" highlighters are mutually exclusive, I believed that the modal ones (alignment column, surrounding quotes & brackets) would continue to operate.

When you are using S/R tool, the "spotlight" highlight comes into play but does not overwrite the other activities (although its highlight should be on top if there is a conflict).

"Remove highlights" only removes manual highlights"

## What actually happens

The "manual" highlighters do seem to operate as I described, removing highlights before applying the new ones. However, they also appear to **turn off** the modal highlighters too. This is what I didn't realize.

The "spotlight" highlight from S/R does interplay properly with the manual highlighters - and takes precedence in case of a conflict. So if you highlighted manually "conductor" and then search "conduct" you would end up with "conduct" spotlight-highlighted, and "or" at the end is in the manual highlight color. If you move on to the next match the "conductor" highlight remains in place. Similar for "highlight single quotes" and "highlight double", it interplays cleanly.

S/R also does not interfere with the modal highlighters.

"Remove highlights" actually removes *all* highlights:
- turns off alignment column
- turns off surrounding quotes/brackets
- removes any manual highlight
- even removes spotlights like matches in S/R

## Questions

- Do we want the manual highlighters (single-quote, double-quote, string-or-regex) to continue to be mutually exclusive, only one in play at a time? I believe **yes**, this is the only way that the highlighting makes sense.
- Do we want the modal highlighters to continue to be disabled, or should they stay active even when manually highlighting?
- If we want them to persist/stay on, do we want the "manual" highlights to take precedence ("win"), where there is a conflict? I would suggest **yes** for this one.
- Should "remove highlights" remove *everything* as in the GG1 behavior?
- Should "remove highlights" move back to the menu position in GG1? Recall that we had moved it higher - but in light of how GG1's remove highlights actually works, the original position makes some more sense than I previously believed.

So in other words do we want to match GG1's behavior, or modify it?

**Comment by windymilla** [comment]
> * Do we want the manual highlighters (single-quote, double-quote, string-or-regex) to continue to be mutually exclusive, only one in play at a time? I believe **yes**, this is the only way that the highlighting makes sense.

Yes
> * Do we want the modal highlighters to continue to be disabled, or should they stay active even when manually highlighting?

I think stay active
> * If we want them to persist/stay on, do we want the "manual" highlights to take precedence ("win"), where there is a conflict? I would suggest **yes** for this one.

Yes
> * Should "remove highlights" remove _everything_ as in the GG1 behavior?

I think no - just remove "manual" highlights. The modal highlighters should be turned on/off via their own menu entry
> * Should "remove highlights" move back to the menu position in GG1? Recall that we had moved it higher - but in light of how GG1's remove highlights actually works, the original position makes some more sense than I previously believed

No - because of the answer to the previous question

Those are my thoughts anyway - what do you think @srjfoo?

**Comment by tangledhelix** [comment]
I'm also wondering if I should create a highlight ranker, run in `on_change`, which runs after the rest of the code and ranks the highlights, in this ranking, so far. This would be running a few `.raise()` or `.lower()` with appropriate args, which I hope wouldn't be a complicated mess.

- lowest: cursorline (which is also suppressed when a selection is active)
- next: align column
- next: surrounding quotes/brackets
- next: manual highlights (single/double quotes, string/regex)
- next/top: spotlight (e.g. current match of a search)

I'm not sue if one can raise/lower the selection highlight (probably we could?) If that's possible, then I propose it belongs between "surrounding quotes/brackets" and "manual highlights". Rationale:

- it has to be below manual highlights since those apply **to** the selected text and we'd want to be able to run that highlight and leave text selected while also displaying it.
- IMO it should go above surrounding quotes/brackets because those are based on the insertion point which will move when a selection is in play. In fact, perhaps that highlight should not even run if a selection is active...

I'll make a separate ticket for this ranking idea unless y'all think I shouldn't do it.

**Comment by windymilla** [comment]
This actually relates to a point I made previously (and said we don't need to think about for now, but perhaps we should).
There should be no need to rank the highlights in `on_change`. What we should be doing is creating the tags once - probably when the MainText is created (and rank them at that point). Then if we are doing highlighting, we just use the pre-existing tags. When we want to clear highlighting, instead of deleting the tag, we can use something like `tag_remove(tagname, "1.0", "end")` which removes the application of the tag to the text without deleting the tag (and losing its ranking).

Advantages of working this way:
1. Single setup during main text creation
2. Simpler highlighting code
3. Less stuff to be done in `on_change` which is called many times
4. Avoids anything like memory leaks in repeated tag create/delete every time `on_change` is called.

> I'm not sure if one can raise/lower the selection highlight (probably we could?) If that's possible, then I propose it belongs between "surrounding quotes/brackets" and "manual highlights"

I think it should be possible - the only thing you can't do with the `sel` tag is delete it, AFAIK.

By the way (& slightly off topic), a cool feature I've never used is binding mouse clicks to tagged text. I'm not sure we have a need for it, but you could envisage a future enhancement where you could either click a bit of tagged text to correct it, or maybe have a customized right-click/context menu pop up when tagged text is clicked. Just thought I'd mention it, in case anyone thought of  a place where we might want it.

**Comment by tangledhelix** [comment]
> This actually relates to a point I made previously (and said we don't need to think about for now, but perhaps we should). There should be no need to rank the highlights in `on_change`. What we should be doing is creating the tags once - probably when the MainText is created (and rank them at that point). Then if we are doing highlighting, we just use the pre-existing tags. When we want to clear highlighting, instead of deleting the tag, we can use something like `tag_remove(tagname, "1.0", "end")` which removes the application of the tag to the text without deleting the tag (and losing its ranking).

Makes sense, and now I recall your earlier point. I'll open a ticket to look at refactoring to do this, before finishing the manual highlight implementation. I think it will simplify some future work too. There won't be any need to configure or figure out WHEN to configure... I got stuck on this ticket for a while on that exact problem yesterday. I thought my highlights weren't applying properly, when it was simply that they were applying an unconfigured tag (which is invisible by default).

**Comment by windymilla** [comment]
I think the only way to create a tag is to apply it to something, but it ought to be possible when the MainText is created to do something like this for each tag (highest rank first):
```
self.tag_add("mytag", "end") # or maybe "1.0" if "end" doesn't work
self.tag_lower("mytag") # puts it at the bottom
self.tag_remove("mytag", "1.0", "end")
```
That would create the tag and apply it to the end/start of the file, then make it the lowest ranked tag, then remove the tag from the whole text, but keep it as an object.
After creating all the tags, something like this should position "sel" below "manual_highlights":
```
self.tag_lower("sel", below_this="manual_highlights")
```

**Comment by srjfoo** [comment]
> > * Do we want the manual highlighters (single-quote, double-quote, string-or-regex) to continue to be mutually exclusive, only one in play at a time? I believe **yes**, this is the only way that the highlighting makes sense.
> 
> Yes
> 
> > * Do we want the modal highlighters to continue to be disabled, or should they stay active even when manually highlighting?
> 
> I think stay active
> 
> > * If we want them to persist/stay on, do we want the "manual" highlights to take precedence ("win"), where there is a conflict? I would suggest **yes** for this one.
> 
> Yes
> 
> > * Should "remove highlights" remove _everything_ as in the GG1 behavior?
> 
> I think no - just remove "manual" highlights. The modal highlighters should be turned on/off via their own menu entry
> 
> > * Should "remove highlights" move back to the menu position in GG1? Recall that we had moved it higher - but in light of how GG1's remove highlights actually works, the original position makes some more sense than I previously believed
> 
> No - because of the answer to the previous question
> 
> Those are my thoughts anyway - what do you think @srjfoo?

I'm good with your choices and reasoning with the caveat that since it's a change from GG1 behavior, it should be called out as such.
